### PR TITLE
HPCC-17447 Document built in compile time constants

### PR DESCRIPTION
**Pull request by JamesDeFabia**
Signed-off-by: Jim DeFabia <jamesdefabia@lexisnexis.com>

@ghalliday Please review for accuracy
@g-pan Please review for formatting, etc.

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

### CODE REVIEW

File: docs/ECLLanguageReference/ECLR_mods/Basics-Constants.xml [hunk 1]
```diff
@@ -152,4 +152,100 @@ MyReal1 := 10.0;   // value of MyReal1 is the REAL value 10.0
...
+  #ELSE
+    FUNCTION
+      string14 fGetDimeTime():= // 14 characters returned
+      BEGINC++
+      #option action
+      struct tm localt;         // localtime in "tm" structure
+      time_t timeinsecs;        // variable to store time in secs
+      time(&amp;timeinsecs);  
+      localtime_r(&amp;timeinsecs,&amp;localt);
+      strftime(__result, 15, "%Y%m%d%H%M%S", &amp;localt); // Formats the localtime to YYYYMMDDhhmmss
```
Comment (ghalliday) [inline docs/ECLLanguageReference/ECLR_mods/Basics-Constants.xml]: Bad code - this corrupts memory.  It needs to write to a temporary, and then copy 14 bytes from it.

**Review by ghalliday — CHANGES_REQUESTED** [review]
Documentation looks good except for the example.

**Comment by JamesDeFabia** [comment]
@ghalliday Edits made, back to you

File: docs/ECLLanguageReference/ECLR_mods/Basics-Constants.xml [hunk 2]
```diff
@@ -152,4 +152,102 @@ MyReal1 := 10.0;   // value of MyReal1 is the REAL value 10.0
...
+              <entry>A BOOLEAN value indicating if it is being compiled to a
+              stand-alone executable.</entry>
+            </row>
+          </tbody>
+        </tgroup>
+      </informaltable>Example:</para>
+
+    <para><programlisting>IMPORT STD;
+  STRING14 fGetDateTimeString() :=
+  #IF(__ECL_VERSION__ &gt;= '5.2')
```
Comment (ghalliday) [inline docs/ECLLanguageReference/ECLR_mods/Basics-Constants.xml]: I seem to have lost a comment on this line.  This doesn't work once the major version goes above 9.  Should be

(__ECL_VERSION_MAJOR__ > 5) or ((__ECL_VERSION_MAJOR__ = 5) AND (__ECL_VERSION_MINOR__ >= 2))

(untested)

File: docs/ECLLanguageReference/ECLR_mods/Basics-Constants.xml [hunk 3]
```diff
@@ -152,4 +152,102 @@ MyReal1 := 10.0;   // value of MyReal1 is the REAL value 10.0
...
+
+    <para><informaltable colsep="1" frame="all" rowsep="1">
+        <tgroup cols="2">
+          <colspec colwidth="150pt" />
+
+          <colspec />
+
+          <tbody>
+            <row>
+              <entry><emphasis>_ECL_VERSION_<indexterm>
```
Comment (ghalliday) [inline docs/ECLLanguageReference/ECLR_mods/Basics-Constants.xml]: All these symbols should be prefixed and suffixed with double underscores.

**Review by ghalliday — CHANGES_REQUESTED** [review]
Sorry, one comment from 1st review got lost, and spotted something else on second view.

**Comment by ghalliday** [comment]
@JamesDeFabia

**Comment by JamesDeFabia** [comment]
@ghalliday More edits made, back to you

**Comment by ghalliday** [comment]
@JamesDeFabia should you have also fixed <primary> tags.
Also, please can you squash the commits.

File: docs/ECLLanguageReference/ECLR_mods/Basics-Constants.xml [hunk 4]
```diff
@@ -171,7 +171,7 @@ MyReal2 := 1.0e1;  // value of MyReal2 is the REAL value 10.0
 
           <tbody>
             <row>
-              <entry><emphasis>_ECL_VERSION_<indexterm>
+              <entry><emphasis>__ECL_VERSION_<indexterm>
```
Comment (richardkchapman) [inline docs/ECLLanguageReference/ECLR_mods/Basics-Constants.xml]: trailing is a double underbar too, I think (and on the others)

**Comment by JamesDeFabia** [comment]
@ghalliday Final edits made and squashed

**Comment by JamesDeFabia** [comment]
@ghalliday  
wait, RKC commented... back to editing

**Comment by JamesDeFabia** [comment]
@ghalliday OK, all edits made and squashed.. back to you

**Review by ghalliday — APPROVED** [review]

**Pull request merged**